### PR TITLE
cdc: Add an "end-of-record" column to

### LINKED ISSUE #7435: CDC log should include an end of record marker
Since a CDC "event" is broken up into potentially several log rows, each representing a sub-modification or pre/post state of a given change/source mutation, a reader cannot really be sure if he has actually read a full event unless he explicitly sees the next timestamp set being. 

Combined with paging and query limits this makes it more difficult to implement a CDC reader/translator without resorting to over-reads and/or stateful caching between calls. 

By adding an end-of-row marker to the last row in a given timestamp group we can always know if we are in fact at an event boundary, even if we don't have "row spill" to check, which makes it easier to write a forward-only reader/translator.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #7435

Adds an "eor" (end-of-record) column to cdc log. This is non-null only on
last-in-timestamp group rows, i.e. end of a singular source "event".

A client can use this as a shortcut to knowing whether or not he has a
full cdc "record" for a given source mutation (single row change).

### CODE REVIEW

File: cdc/log.cc [hunk 1]
```diff
@@ -519,6 +519,7 @@ static schema_ptr create_log_schema(const schema& s, std::optional<utils::UUID>
     b.with_column(log_meta_column_name_bytes("batch_seq_no"), int32_type, column_kind::clustering_key);
     b.with_column(log_meta_column_name_bytes("operation"), data_type_for<operation_native_type>());
     b.with_column(log_meta_column_name_bytes("ttl"), long_type);
+    b.with_column(log_meta_column_name_bytes("eor"), boolean_type);
```
[the remark this answers is not in the document]
Comment (elcallio) [reply cdc/log.cc:522]: I purposefully picked as short a column name as I could, but if you want to make it more verbatim, how about "end_of_batch"?

File: test/boost/cdc_test.cc [hunk 2]
```diff
@@ -369,7 +370,7 @@ SEASTAR_THREAD_TEST_CASE(test_cdc_log_schema) {
         assert_has_column(cdc::log_data_column_deleted_elements_name("c_typ"), set_type_impl::get_instance(short_type, false));
 
         // Check if we missed something
-        BOOST_REQUIRE_EQUAL(required_column_count, log_schema->all_columns_count());
+        BOOST_REQUIRE_GE(required_column_count, log_schema->all_columns_count());
```
Comment (haaawk) [inline test/boost/cdc_test.cc:373]: Why is this needed?
assert_has_column(cdc::log_meta_column_name("eor"), boolean_type); should have increased required_column_count by 1, no?

Comment (elcallio) [reply test/boost/cdc_test.cc:373]: It is not, but I found the check pernickety. Have _more_ columns than those required should not really be a problem.

File: cdc/log.cc [hunk 3]
```diff
@@ -944,6 +948,11 @@ class log_mutation_builder {
         _log_mut.set_cell(log_ck, log_cdef, atomic_cell::make_live(*log_cdef.type, _ts, deleted_elements, _ttl));
     }
 
+    void end_record() {
```
[the remark this answers is not in the document]
Comment (elcallio) [reply cdc/log.cc:951]: ? Look at the call chain. the caller (split.cc) only has _source_ ck.

File: cdc/log.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (haaawk) [reply cdc/log.cc:522]: end of batch is fine

File: test/boost/cdc_test.cc (hunk 2 above)
Comment (haaawk) [reply test/boost/cdc_test.cc:373]: I think it's better to be precise. Otherwise we may miss errors when one column is added but tests are not updated and then another column is added and tests are just updated by +1. Then the test wouldn't catch the second column not working.

File: cdc/log.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (haaawk) [reply cdc/log.cc:951]: Ok. I see this is needed.

**Review by haaawk — APPROVED** [review]

File: cdc/log.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (elcallio) [reply cdc/log.cc]: Done

File: test/boost/cdc_test.cc (hunk 2 above)
Comment (elcallio) [reply test/boost/cdc_test.cc]: Reverted

**Review by kbr- — APPROVED** [review]

**Review by kbr- — COMMENTED** [review]
wait, the new version seems to do some new stuff, lemme review again

**Comment by elcallio** [comment]
I changed the impl a tiny bit to requests from  @haaawk

File: cdc/log.cc [hunk 4]
```diff
@@ -944,8 +957,13 @@ class log_mutation_builder {
...
+        if (has_rows()) {
+            _log_mut.set_cell(last_row_key(), log_meta_column_name_bytes("end_of_batch"), data_value(true), _ts, _ttl);
+        }
+    }
 private:
-    void set_key_columns(const clustering_key& log_ck, schema::const_iterator_range_type columns, const std::vector<bytes>& key) {
+    void set_key_columns(const clustering_key& log_ck, schema::const_iterator_range_type columns, const std::vector<bytes>& key) const {
```
Comment (kbr-) [inline cdc/log.cc:966]: But it's not const, it modifies the state.

File: cdc/log.cc [hunk 5]
```diff
@@ -880,12 +881,24 @@ class log_mutation_builder {
         return _base_schema;
     }
 
+    clustering_key create_ck(int batch) const {
+        auto log_ck = clustering_key::from_exploded(_log_schema, { _tuuid, int32_type->decompose(batch) });
+        set_key_columns(log_ck, _base_schema.partition_key_columns(), _base_pk);
```
Comment (kbr-) [inline cdc/log.cc:886]: this is definitely not `const`.
I see what you wanted to do here, leme suggest the following: move the `set_key_columns` call to `allocate_new_log_row`. Then `create_ck` is indeed `const` (it's just a calculation), and `set_key_columns` doesn't need to be `const`, as it should not be.

File: cdc/log.cc [hunk 6]
```diff
@@ -880,12 +881,24 @@ class log_mutation_builder {
...
-        set_key_columns(log_ck, _base_schema.partition_key_columns(), _base_pk);
-        return log_ck;
+        return create_ck(_batch_no++);
+    }
+
+    bool has_rows() const {
+        return _batch_no != 0;
+    }
+
+    clustering_key last_row_key() const {
```
Comment (kbr-) [inline cdc/log.cc:900]: One would presume that this is a pure calculation, but it actually modifies the state now.

File: cdc/log.cc (hunk 5 above)
Comment (elcallio) [reply cdc/log.cc:886]: true.

File: cdc/log.cc (hunk 6 above)
Comment (elcallio) [reply cdc/log.cc:900]: Fixed.

File: cdc/log.cc (hunk 5 above)
Comment (elcallio) [reply cdc/log.cc]: Fixed.

**Review by kbr- — APPROVED** [review]

**Comment by haaawk** [comment]
@nyh I think the PR is ready for you to review and merge if you find it appropriate. Thanks.

**Comment by elcallio** [comment]
@nyh ping?
